egs/aishell/asr/simple_v1/mmi_bigram_decode.py

```diff
@@ -0,0 +1,306 @@
+    return k2.ragged.to_list(aux_labels)
+
+
+def invert_permutation(indices: torch.Tensor) -> torch.Tensor:
+    ans = torch.zeros(indices.shape, device=indices.device, dtype=torch.long)
+    ans[indices] = torch.arange(0, indices.shape[0], device=indices.device)
+    return ans
+
+def find_first_disambig_symbol(symbols: SymbolTable) -> int:
+    return min(v for k, v in symbols._sym2id.items() if k.startswith('#'))
+
+
+def print_transition_probabilities(P: k2.Fsa, phone_symbol_table: SymbolTable,
+                                   phone_ids: List[int], filename: str):
+    '''Print the transition probabilities of a phone LM.
+
+    Args:
+      P:
+        A bigram phone LM.
+      phone_symbol_table:
+        The phone symbol table.
+      phone_ids:
+        A list of phone ids
+      filename:
+        Filename to save the printed result.
+    '''
+    num_phones = len(phone_ids)
+    table = np.zeros((num_phones + 1, num_phones + 2))
+    table[:, 0] = 0
+    table[0, -1] = 0 # the start state has no arcs to the final state
+    assert P.arcs.dim0() == num_phones + 2
+    arcs = P.arcs.values()[:, :3]
+    probability = P.scores.exp().tolist()
+
+    assert arcs.shape[0] - num_phones == num_phones * (num_phones + 1)
+    for i, arc in enumerate(arcs.tolist()):
+        src_state, dest_state, label = arc[0], arc[1], arc[2]
+        prob = probability[i]
+        if label != -1:
+            assert label == dest_state
+        else:
+            assert dest_state == num_phones + 1
+        table[src_state][dest_state] = prob
+
+    try:
+        from prettytable import PrettyTable
+    except ImportError:
+        print('Please run `pip install prettytable`. Skip printing')
+        return
+
+    x = PrettyTable()
+
+    field_names = ['source']
+    field_names.append('sum')
+    for i in phone_ids:
+        field_names.append(phone_symbol_table[i])
+    field_names.append('final')
+
+    x.field_names = field_names
+
+    for row in range(num_phones + 1):
+        this_row = []
+        if row == 0:
+            this_row.append('start')
+        else:
+            this_row.append(phone_symbol_table[row])
+        this_row.append('{:.6f}'.format(table[row, 1:].sum()))
+        for col in range(1, num_phones + 2):
+            this_row.append('{:.6f}'.format(table[row, col]))
+        x.add_row(this_row)
+    with open(filename, 'w') as f:
+        f.write(str(x))
+
+
+def main():
+    exp_dir = Path('exp-lstm-adam-mmi-bigram-musan')
+    setup_logger('{}/log/log-decode'.format(exp_dir), log_level='debug')
+
+    # load L, G, symbol_table
+    lang_dir = Path('data/lang_nosp')
+    symbol_table = k2.SymbolTable.from_file(lang_dir / 'words.txt')
+    phone_symbol_table = k2.SymbolTable.from_file(lang_dir / 'phones.txt')
+
+    phone_ids = get_phone_symbols(phone_symbol_table)
+    P = create_bigram_phone_lm(phone_ids)
+
+    phone_ids_with_blank = [0] + phone_ids
+    ctc_topo = build_ctc_topo(phone_ids_with_blank)
```

Comment:
The latest master branch has changed the return value of `build_ctc_topo`.
There is no need to call `ctc_topo.invert()` now.

Please have a look at
https://github.com/k2-fsa/snowfall/blob/178fdee5d5d5cc212130922f530d5f8363f38a85/egs/librispeech/asr/simple_v1/mmi_bigram_decode.py#L210

and

https://github.com/k2-fsa/snowfall/blob/178fdee5d5d5cc212130922f530d5f8363f38a85/egs/librispeech/asr/simple_v1/mmi_bigram_decode.py#L242-L246

Also, the function argument for `compile_LG` has been changed.